snuba/clusters/__init__.py

```diff
@@ -0,0 +1,54 @@
+from snuba import settings
+
+from snuba.clickhouse.native import ClickhousePool, NativeDriverReader
+from snuba.clickhouse.query import ClickhouseQuery
+from snuba.reader import Reader
+
+class Cluster:
+    """
+    A cluster is responsible for managing a collection of database nodes.
+
+    Clusters are configurable, and will be instantiated based on user defined settings.
+
+    Each storage must be mapped to a cluster, and this mapping is configurable
+    (with the constraint that storages that form part of joined queries need to be
+    located on the same cluster).
+
+    A cluster provides a reader and Clickhouse connections that are shared by all
+    storages located on the cluster.
+
+    In future, clusters will also be reponsible for co-ordinating commands that
+    need to be run on multiple hosts that are colocated within the same cluster -
+    e.g. bootstrap, migration, cleanup, optimize.
+    """
+    def __init__(self, host: str, port: int, http_port: int):
+        self.__clickhouse_rw = ClickhousePool(host, port)
+        self.__clickhouse_ro = ClickhousePool(
+            host,
+            port,
+            client_settings={"readonly": True},
+        )
+        self.__reader: Reader[ClickhouseQuery] = NativeDriverReader(self.__clickhouse_ro)
+
+    def get_clickhouse_rw(self) -> ClickhousePool:
+        return self.__clickhouse_rw
+
+    def get_clickhouse_ro(self) -> ClickhousePool:
+        return self.__clickhouse_ro
+
+    def get_reader(self) -> Reader:
+        return self.__reader
+
+
+cluster = Cluster(
+    host=settings.CLICKHOUSE_HOST,
+    port=settings.CLICKHOUSE_PORT,
+    http_port=settings.CLICKHOUSE_HTTP_PORT,
+)
```

Comment:
This works here, but I think trying to consolidate these resources in `snuba.environment` will be helpful over the long term because it enables us to eventually consolidate all of the different object instances that need to have some sort of lifecycle control.

There is a tech debt ticket for this, and [some more details here](https://github.com/getsentry/snuba/pull/772#discussion_r377381301), but I guess as of right now this is just a personal preference and not anything we've really discussed the overall merits and direction of… maybe we should?

Reply:
ping

Reply:
Agree we should decide on a consistent approach, can we hold off for now? Most other resources are not handled in that way yet. There's a bit more going on in here now and I'd prefer to keep it centralised in clusters until we make a decision.